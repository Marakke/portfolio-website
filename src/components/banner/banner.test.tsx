import React from "react";
import { render, fireEvent } from "@testing-library/react";
import * as Gatsby from "gatsby";
import Banner from "./index";

const useStaticQuery = jest.spyOn(Gatsby, `useStaticQuery`);

const mockUseStaticQuery = {
  banner: {
    text: "Banner Text",
    audio: {
      url: "mock/audio/path.mp3",
    },
    buttons: [ 
      {
        logo: {
          fluid() {
            return {
              aspectRatio: 1,
              src: "mock/image/path.jpg",
              srcSet: "mock/srcSet",
              sizes: "(max-width: 800px) 100vw, 800px",
            };
          },
          alt: "Test Alt Text"
        },
        name: "Button Name",
        link: "Button Link"
      }
    ] 
  },
};

describe("Banner", () => {
  beforeEach(() => {
    useStaticQuery.mockImplementation(() => mockUseStaticQuery);
    window.open = jest.fn();
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("renders the Banner component correctly", () => {
    const { container, getByTestId, getByAltText, getByRole } = render(<Banner />);
    expect(container).toMatchSnapshot();

    const bannerContainer = getByTestId("banner-container");
    expect(bannerContainer).toBeInTheDocument();

    const textElement = getByRole("heading", { name: "Banner Text" });
    expect(textElement).toBeInTheDocument();

    const imageElement = getByAltText("Test Alt Text");
    expect(imageElement).toBeInTheDocument();

    const description = container.querySelectorAll("p");
    expect(description[0].textContent).toBe("Button Name");

    const audioElement = container.querySelector("audio");
    expect(audioElement).toBeInTheDocument();
    expect(audioElement).toHaveAttribute("src", "mock/audio/path.mp3");
    expect(audioElement).toHaveAttribute("controls");
    expect(audioElement).toHaveAttribute("controlsList", "nodownload noplaybackrate");
  });

  it("clicks banner button and opens new tab", () => {
    const { container } = render(<Banner />);

    const buttonElement = container.querySelectorAll("button");
    expect(buttonElement[0]).toBeInTheDocument();

    fireEvent.click(buttonElement[0]);

    expect(window.open).toHaveBeenCalledWith("Button Link", "_blank");
  });
});
