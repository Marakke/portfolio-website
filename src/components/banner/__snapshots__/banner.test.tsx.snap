// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner renders the Banner component correctly 1`] = `
<div>
  <div
    class="banner-container"
    data-testid="banner-container"
  >
    <audio
      controls=""
      controlslist="nodownload noplaybackrate"
      src="mock/audio/path.mp3"
    />
    <h2
      class="banner-text"
    >
      Banner Text
    </h2>
    <div
      class="banner-buttons"
    >
      <button
        class="banner-button"
      >
        <div
          class="logo gatsby-image-wrapper"
          style="position: relative; overflow: hidden;"
        >
          <div
            aria-hidden="true"
            style="width: 100%;"
          />
          <picture>
            <img
              alt="Test Alt Text"
              loading="lazy"
              style="position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; object-fit: cover; object-position: center; opacity: 0; transition: opacity 500ms;"
            />
          </picture>
          <noscript>
            &lt;picture&gt;&lt;source srcset="undefined" /&gt;&lt;img loading="lazy" src="" alt="Test Alt Text" style="position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center"/&gt;&lt;/picture&gt;
          </noscript>
        </div>
        <p
          class="banner-button-text"
        >
          Button Name
        </p>
      </button>
    </div>
  </div>
</div>
`;
